File: src/graphics/cpp/transformations/main_window.cpp
#include "main_window.h"
#include "./ui_main_window.h"

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow) {
  ui->setupUi(this);

  ui->xSpinBox->setValue(ui->transformationDisplay->getTranslate().x());
  ui->ySpinBox->setValue(ui->transformationDisplay->getTranslate().y());
  ui->scaleSpinBox->setValue(ui->transformationDisplay->getScale() * 100);
  ui->rotateSpinBox->setValue(ui->transformationDisplay->getRotate());

  // conect ui->transformationDisplay setTranslate() to xSpinBox and ySpinBox
  connect(ui->xSpinBox, &QSpinBox::valueChanged, ui->transformationDisplay,
          [&](int x) {
            ui->transformationDisplay->setTranslate(
                x, ui->transformationDisplay->getTranslate().y());
          });

  connect(ui->ySpinBox, &QSpinBox::valueChanged, ui->transformationDisplay,
          [&](int y) {
            ui->transformationDisplay->setTranslate(
                ui->transformationDisplay->getTranslate().x(), y);
          });

  // conect ui->transformationDisplay setScale() to scaleSpinBox
  connect(ui->scaleSpinBox, &QDoubleSpinBox::valueChanged,
          ui->transformationDisplay, [&](double scale) {
            ui->transformationDisplay->setScale(scale / 100);
          });

  // conect ui->transformationDisplay setRotate() to rotateSpinBox
  connect(ui->rotateSpinBox, &QSpinBox::valueChanged, ui->transformationDisplay,
          [&](int angle) { ui->transformationDisplay->setRotate(angle); });
}

MainWindow::~MainWindow() { delete ui; }

File: src/graphics/python/transformations/main_window.py
from PyQt6 import QtWidgets, uic, QtGui
from PyQt6.QtCore import QPoint


class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        uic.loadUi("main_window.ui", self)
        self.xSpinBox.setValue(self.transformationDisplay.translate.x())
        self.ySpinBox.setValue(self.transformationDisplay.translate.y())
        self.scaleSpinBox.setValue(self.transformationDisplay.scale * 100)
        self.rotateSpinBox.setValue(self.transformationDisplay.rotate)

        # connect ui->transformationDisplay setTranslate() to xSpinBox and ySpinBox
        self.xSpinBox.valueChanged.connect(
            lambda x: self.transformationDisplay.update_translate(
                QPoint(x, self.transformationDisplay.translate.y())
            )
        )
        self.ySpinBox.valueChanged.connect(
            lambda y: self.transformationDisplay.update_translate(
                QPoint(self.transformationDisplay.translate.x(), y)
            )
        )
        self.scaleSpinBox.valueChanged.connect(
            lambda scale: self.transformationDisplay.update_scale(scale / 100)
        )
        self.rotateSpinBox.valueChanged.connect(
            lambda angle: self.transformationDisplay.update_rotate(angle)
        )

        self.show()
